File: cpp_src/puppy_core/robot_state_aggregator.cpp
// Robot State Aggregator 实现
#include "puppy_core/robot_state_aggregator.h"

#include <algorithm>
#include <set>
#include <sstream>

namespace puppy_core {

// 默认超时阈值（秒）
static constexpr double kDefaultBatteryTimeout = 10.0;
static constexpr double kDefaultMotionTimeout = 5.0;
static constexpr double kDefaultImuTimeout = 2.0;
static constexpr double kDefaultLidarTimeout = 3.0;

RobotStateAggregatorNode::RobotStateAggregatorNode()
    : rclcpp::Node("robot_state_aggregator") {
  this->declare_parameter("battery_timeout", kDefaultBatteryTimeout);
  this->declare_parameter("motion_timeout", kDefaultMotionTimeout);
  this->declare_parameter("imu_timeout", kDefaultImuTimeout);
  this->declare_parameter("lidar_timeout", kDefaultLidarTimeout);
  battery_timeout_ = this->get_parameter("battery_timeout").as_double();
  motion_timeout_ = this->get_parameter("motion_timeout").as_double();
  imu_timeout_ = this->get_parameter("imu_timeout").as_double();
  lidar_timeout_ = this->get_parameter("lidar_timeout").as_double();

  battery_percent_ = 1.0;
  imu_ready_ = false;
  lidar_ready_ = false;
  camera_ready_ = false;
  motion_ready_ = false;
  has_platform_health_ = false;

  // 初始化数据源状态
  for (int i = 0; i < 4; ++i) {
    Source s = static_cast<Source>(i);
    has_seen_[s] = false;
    last_seen_[s] = rclcpp::Time(0, 0, this->get_clock()->get_clock_type());
  }

  health_pub_ = this->create_publisher<puppy_interfaces::msg::RobotHealth>(
      "/robot/health", 10);

  battery_sub_ =
      this->create_subscription<puppy_interfaces::msg::BatteryStatus>(
          "/battery_status", 10,
          std::bind(&RobotStateAggregatorNode::onBattery, this,
                    std::placeholders::_1));
  motion_sub_ =
      this->create_subscription<puppy_interfaces::msg::PlatformMotionState>(
          "/platform/motion_state", 10,
          std::bind(&RobotStateAggregatorNode::onMotion, this,
                    std::placeholders::_1));
  fall_sub_ = this->create_subscription<puppy_interfaces::msg::FallEvent>(
      "/fall/event", 10,
      std::bind(&RobotStateAggregatorNode::onFall, this,
                std::placeholders::_1));
  platform_health_sub_ =
      this->create_subscription<puppy_interfaces::msg::RobotHealth>(
          "/platform/health", 10,
          std::bind(&RobotStateAggregatorNode::onPlatformHealth, this,
                    std::placeholders::_1));
  imu_sub_ = this->create_subscription<sensor_msgs::msg::Imu>(
      "/imu/data", 10,
      std::bind(&RobotStateAggregatorNode::onImu, this,
                std::placeholders::_1));
  scan_sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
      "/scan", 10,
      std::bind(&RobotStateAggregatorNode::onScan, this,
                std::placeholders::_1));

  timer_ = this->create_wall_timer(
      std::chrono::milliseconds(500),
      std::bind(&RobotStateAggregatorNode::publishHealth, this));

  RCLCPP_INFO(this->get_logger(),
              "Robot state aggregator started (timeouts: battery=%.1f, "
              "motion=%.1f, imu=%.1f, lidar=%.1f)",
              battery_timeout_, motion_timeout_, imu_timeout_, lidar_timeout_);
}

const char* RobotStateAggregatorNode::sourceName(Source s) {
  switch (s) {
    case Source::kBattery: return "battery";
    case Source::kMotion:  return "motion";
    case Source::kImu:     return "imu";
    case Source::kLidar:   return "lidar";
  }
  return "unknown";
}

std::string RobotStateAggregatorNode::sourceUpper(Source s) {
  std::string upper = sourceName(s);
  std::transform(upper.begin(), upper.end(), upper.begin(),
                 [](unsigned char c) { return std::toupper(c); });
  return upper;
}

void RobotStateAggregatorNode::onBattery(
    const puppy_interfaces::msg::BatteryStatus::SharedPtr msg) {
  battery_percent_ = std::max(0.0, std::min(1.0, static_cast<double>(msg->percent)));
  has_seen_[Source::kBattery] = true;
  last_seen_[Source::kBattery] = this->now();
}

void RobotStateAggregatorNode::onMotion(
    const puppy_interfaces::msg::PlatformMotionState::SharedPtr msg) {
  motion_ready_ = msg->controllable;
  has_seen_[Source::kMotion] = true;
  last_seen_[Source::kMotion] = this->now();
}

void RobotStateAggregatorNode::onImu(
    const sensor_msgs::msg::Imu::SharedPtr /*msg*/) {
  has_seen_[Source::kImu] = true;
  last_seen_[Source::kImu] = this->now();
}

void RobotStateAggregatorNode::onScan(
    const sensor_msgs::msg::LaserScan::SharedPtr /*msg*/) {
  has_seen_[Source::kLidar] = true;
  last_seen_[Source::kLidar] = this->now();
}

void RobotStateAggregatorNode::onFall(
    const puppy_interfaces::msg::FallEvent::SharedPtr msg) {
  const std::string fault = "FALL_DETECTED";
  if (msg->detected) {
    if (std::find(active_faults_.begin(), active_faults_.end(), fault) ==
        active_faults_.end()) {
      active_faults_.push_back(fault);
    }
  } else {
    auto it = std::find(active_faults_.begin(), active_faults_.end(), fault);
    if (it != active_faults_.end()) {
      active_faults_.erase(it);
    }
  }
}

void RobotStateAggregatorNode::onPlatformHealth(
    const puppy_interfaces::msg::RobotHealth::SharedPtr msg) {
  has_platform_health_ = true;
  platform_health_ = *msg;
  camera_ready_ = msg->camera_ready;
}

std::vector<std::string> RobotStateAggregatorNode::checkTimeouts() {
  rclcpp::Time now = this->now();
  std::vector<std::string> timeout_faults;

  const Source sources[] = {Source::kBattery, Source::kMotion,
                            Source::kImu, Source::kLidar};
  for (Source s : sources) {
    if (!has_seen_[s]) {
      timeout_faults.push_back(sourceUpper(s) + "_OFFLINE");
      continue;
    }
    double elapsed = (now - last_seen_[s]).seconds();
    double threshold = 0.0;
    switch (s) {
      case Source::kBattery: threshold = battery_timeout_; break;
      case Source::kMotion:  threshold = motion_timeout_;  break;
      case Source::kImu:     threshold = imu_timeout_;     break;
      case Source::kLidar:   threshold = lidar_timeout_;   break;
    }
    if (elapsed > threshold) {
      timeout_faults.push_back(sourceUpper(s) + "_TIMEOUT");
    }
  }

  // 新增的故障告警
  for (const auto& fault : timeout_faults) {
    if (reported_timeout_faults_.find(fault) == reported_timeout_faults_.end()) {
      RCLCPP_WARN(this->get_logger(), "Sensor timeout: %s", fault.c_str());
    }
  }

  // 已恢复的故障
  std::set<std::string> current_set(timeout_faults.begin(),
                                    timeout_faults.end());
  for (const auto& fault : reported_timeout_faults_) {
    if (current_set.find(fault) == current_set.end()) {
      RCLCPP_INFO(this->get_logger(), "Sensor recovered: %s", fault.c_str());
    }
  }
  reported_timeout_faults_ = current_set;

  return timeout_faults;
}

void RobotStateAggregatorNode::publishHealth() {
  puppy_interfaces::msg::RobotHealth msg;
  msg.header.stamp = this->now();

  std::vector<std::string> timeout_faults = checkTimeouts();

  // 提取超时故障的源前缀（如 "BATTERY_TIMEOUT" -> "BATTERY"）
  std::set<std::string> timeout_set;
  for (const auto& fault : timeout_faults) {
    auto underscore = fault.find('_');
    if (underscore != std::string::npos) {
      timeout_set.insert(fault.substr(0, underscore));
    } else {
      timeout_set.insert(fault);
    }
  }

  imu_ready_ = (timeout_set.find("IMU") == timeout_set.end()) &&
               has_seen_[Source::kImu];
  lidar_ready_ = (timeout_set.find("LIDAR") == timeout_set.end()) &&
                 has_seen_[Source::kLidar];
  bool battery_ok = (timeout_set.find("BATTERY") == timeout_set.end());
  if (timeout_set.find("MOTION") != timeout_set.end()) {
    motion_ready_ = false;
  }

  std::vector<std::string> active_faults = active_faults_;
  if (has_platform_health_) {
    for (const auto& fault : platform_health_.active_faults) {
      if (std::find(active_faults.begin(), active_faults.end(), fault) ==
          active_faults.end()) {
        active_faults.push_back(fault);
      }
    }
  }
  for (const auto& fault : timeout_faults) {
    if (std::find(active_faults.begin(), active_faults.end(), fault) ==
        active_faults.end()) {
      active_faults.push_back(fault);
    }
  }

  msg.cpu_temp =
      has_platform_health_ ? platform_health_.cpu_temp : 0.0f;
  msg.ok = active_faults.empty() && battery_ok && battery_percent_ > 0.1;
  msg.level = msg.ok ? "OK" : "ERROR";
  msg.active_faults = active_faults;
  msg.battery_percent =
      battery_ok ? static_cast<float>(battery_percent_ * 100.0) : 0.0f;
  msg.imu_ready = imu_ready_;
  msg.lidar_ready = lidar_ready_;
  msg.camera_ready = camera_ready_;
  msg.motion_ready = motion_ready_;
  health_pub_->publish(msg);
}

}  // namespace puppy_core

File: src/puppy_core/puppy_core/robot_state_aggregator.py
"""Robot State Aggregator: unified robot health from semantic topics."""
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Imu, LaserScan

from puppy_interfaces.msg import BatteryStatus, FallEvent, PlatformMotionState, RobotHealth


_DEFAULT_TIMEOUTS = {
    'battery': 10.0,
    'motion': 5.0,
    'imu': 2.0,
    'lidar': 3.0,
}


class RobotStateAggregatorNode(Node):
    """Aggregates robot state from multiple sources into unified health."""

    def __init__(self):
        super().__init__('robot_state_aggregator')

        self.declare_parameter('battery_timeout', _DEFAULT_TIMEOUTS['battery'])
        self.declare_parameter('motion_timeout', _DEFAULT_TIMEOUTS['motion'])
        self.declare_parameter('imu_timeout', _DEFAULT_TIMEOUTS['imu'])
        self.declare_parameter('lidar_timeout', _DEFAULT_TIMEOUTS['lidar'])
        self.timeouts = {
            'battery': self.get_parameter('battery_timeout').value,
            'motion': self.get_parameter('motion_timeout').value,
            'imu': self.get_parameter('imu_timeout').value,
            'lidar': self.get_parameter('lidar_timeout').value,
        }

        self.battery_percent = 1.0
        self.imu_ready = False
        self.lidar_ready = False
        self.camera_ready = False
        self.motion_ready = False
        self.active_faults = []
        self.platform_health = None

        self.last_seen = {
            'battery': None,
            'motion': None,
            'imu': None,
            'lidar': None,
        }
        self._reported_timeout_faults = set()

        self.health_pub = self.create_publisher(RobotHealth, '/robot/health', 10)

        self.create_subscription(BatteryStatus, '/battery_status', self._on_battery, 10)
        self.create_subscription(PlatformMotionState, '/platform/motion_state', self._on_motion, 10)
        self.create_subscription(FallEvent, '/fall/event', self._on_fall, 10)
        self.create_subscription(RobotHealth, '/platform/health', self._on_platform_health, 10)
        self.create_subscription(Imu, '/imu/data', self._on_imu, 10)
        self.create_subscription(LaserScan, '/scan', self._on_scan, 10)

        self.create_timer(0.5, self._publish_health)

        self.get_logger().info(f'Robot state aggregator started (timeouts: {self.timeouts})')

    def _on_battery(self, msg: BatteryStatus):
        self.battery_percent = float(max(0.0, min(1.0, msg.percent)))
        self.last_seen['battery'] = self.get_clock().now()

    def _on_motion(self, msg: PlatformMotionState):
        self.motion_ready = msg.controllable
        self.last_seen['motion'] = self.get_clock().now()

    def _on_imu(self, msg: Imu):
        self.last_seen['imu'] = self.get_clock().now()

    def _on_scan(self, msg: LaserScan):
        self.last_seen['lidar'] = self.get_clock().now()

    def _on_fall(self, msg: FallEvent):
        if msg.detected:
            if 'FALL_DETECTED' not in self.active_faults:
                self.active_faults.append('FALL_DETECTED')
        elif 'FALL_DETECTED' in self.active_faults:
            self.active_faults.remove('FALL_DETECTED')

    def _on_platform_health(self, msg: RobotHealth):
        self.platform_health = msg
        self.camera_ready = msg.camera_ready

    def _check_timeouts(self) -> list:
        now = self.get_clock().now()
        timeout_faults = []
        for source, last in self.last_seen.items():
            if last is None:
                timeout_faults.append(f'{source.upper()}_OFFLINE')
                continue
            elapsed = (now - last).nanoseconds / 1e9
            if elapsed > self.timeouts[source]:
                timeout_faults.append(f'{source.upper()}_TIMEOUT')

        for fault in timeout_faults:
            if fault not in self._reported_timeout_faults:
                self.get_logger().warn(f'Sensor timeout: {fault}')
                self._reported_timeout_faults.add(fault)

        recovered = self._reported_timeout_faults - set(timeout_faults)
        for fault in recovered:
            self.get_logger().info(f'Sensor recovered: {fault}')
        self._reported_timeout_faults = set(timeout_faults)

        return timeout_faults

    def _publish_health(self):
        msg = RobotHealth()
        msg.header.stamp = self.get_clock().now().to_msg()

        timeout_faults = self._check_timeouts()
        timeout_set = {fault.split('_')[0] for fault in timeout_faults}

        self.imu_ready = 'IMU' not in timeout_set and self.last_seen['imu'] is not None
        self.lidar_ready = 'LIDAR' not in timeout_set and self.last_seen['lidar'] is not None
        battery_ok = 'BATTERY' not in timeout_set
        if 'MOTION' in timeout_set:
            self.motion_ready = False

        active_faults = list(self.active_faults)
        if self.platform_health is not None:
            for fault in self.platform_health.active_faults:
                if fault not in active_faults:
                    active_faults.append(fault)
        for fault in timeout_faults:
            if fault not in active_faults:
                active_faults.append(fault)

        msg.cpu_temp = self.platform_health.cpu_temp if self.platform_health else 0.0
        msg.ok = len(active_faults) == 0 and battery_ok and self.battery_percent > 0.1
        msg.level = 'OK' if msg.ok else 'ERROR'
        msg.active_faults = active_faults
        msg.battery_percent = (self.battery_percent * 100.0) if battery_ok else 0.0
        msg.imu_ready = self.imu_ready
        msg.lidar_ready = self.lidar_ready
        msg.camera_ready = self.camera_ready
        msg.motion_ready = self.motion_ready
        self.health_pub.publish(msg)


def main(args=None):
    rclpy.init(args=args)
    node = RobotStateAggregatorNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()
